In [7]:
import numpy as np
P = np.array([[0,0.4,0.2,0.4],[0.5, 0.5,0,0],[1,0,0,0],[1,0,0,0]])
p0 = np.ones(4)/4
p10 = p0 @ np.linalg.matrix_power(P, 10)
print("P**10", np.linalg.matrix_power(P, 10))
p10

P**10 [[0.52088625 0.30249625 0.0588725  0.117745  ]
 [0.37812031 0.34475281 0.09237563 0.18475125]
 [0.2943625  0.3695025  0.112045   0.22409   ]
 [0.2943625  0.3695025  0.112045   0.22409   ]]


array([0.37193289, 0.34656352, 0.09383453, 0.18766906])

# 1.5

In [13]:
np.random.seed(42)
def simulate_chain(P, p0, N):
    states = np.zeros(N, dtype=int)
    states[0] = np.random.choice(4, p=p0)

    for n in range(1, N):
        states[n] = np.random.choice(4, p=P[states[n-1]])

    return states
def empirical_distribution(states, n):
    counts = np.bincount(states[:n], minlength=4)
    return counts / n
N = 100
print(f"N={N}")
states = simulate_chain(P, p0, N)

# empirical distributions at even and odd times
p_even = empirical_distribution(states[::2], N // 2)
p_odd  = empirical_distribution(states[1::2], N // 2)

print("Empirical distribution (even times):", p_even)
print("Empirical distribution (odd times): ", p_odd)

N = 1000000
print(f"N={N}")
states = simulate_chain(P, p0, N)

# empirical distributions at even and odd times
p_even = empirical_distribution(states[::2], N // 2)
p_odd  = empirical_distribution(states[1::2], N // 2)

print("Empirical distribution (even times):", p_even)
print("Empirical distribution (odd times): ", p_odd)



N=100
Empirical distribution (even times): [0.48 0.46 0.04 0.02]
Empirical distribution (odd times):  [0.24 0.56 0.04 0.16]
N=1000000
Empirical distribution (even times): [0.417396 0.332328 0.083168 0.167108]
Empirical distribution (odd times):  [0.416444 0.332076 0.0837   0.16778 ]


The distribution appears to converge to a stationary distribution with approximate probability 0.41 of being at an intersection, probability 0.33 of going forward, 0.083 of turning left and 0.17 of turning right. We expect this, since we have an irreducible chain (every state can reach every other state), finite number of states, and aperiodicity (the states have a self-loop: they can return to themslves in 1 step).